runner/local/simple.go

```diff
@@ -2,69 +2,136 @@ package local
-	mu        sync.Mutex
+	exec       execer.Execer
+	runs       map[runner.RunId]runner.ProcessStatus
+	running    map[runner.RunId]execer.Process
+	maxRunning int
+	nextRunId  int64
+	mu         sync.Mutex
 }
 
-func (r *simpleRunner) Run(cmd *runner.Command) (runner.ProcessStatus, error) {
+func (r *simpleRunner) Run(cmd *runner.Command) runner.ProcessStatus {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	runId := runner.RunId(fmt.Sprintf("%d", r.nextRunId))
 	r.nextRunId++
 
-	if r.running != nil {
+	if len(r.running) >= r.maxRunning {
 		r.runs[runId] = runner.ErrorStatus(runId, fmt.Errorf("Runner is busy"))
-		return r.runs[runId], nil
+		return r.runs[runId]
 	}
 
 	var c execer.Command
 	c.Argv = cmd.Argv
 	p, err := r.exec.Exec(c)
 	if err != nil {
 		r.runs[runId] = runner.ErrorStatus(runId, fmt.Errorf("could not exec: %v", err))
-		return r.runs[runId], err
+		return r.runs[runId]
 	}
 	r.runs[runId] = runner.RunningStatus(runId)
-	r.running = p
+	r.running[runId] = p
 
-	go babysit(p, runId, r)
+	go babysit(p, runId, r, cmd.Timeout)
 
-	return r.runs[runId], nil
+	return r.runs[runId]
 }
 
-func (r *simpleRunner) Status(run runner.RunId) (runner.ProcessStatus, error) {
+func (r *simpleRunner) Status(run runner.RunId) runner.ProcessStatus {
 	r.mu.Lock()
 	defer r.mu.Unlock()
 	result, ok := r.runs[run]
 	if !ok {
-		return result, fmt.Errorf("unknown runId")
+		return runner.ErrorStatus(run, fmt.Errorf("could not find: %v", run))
+	}
+	return result
+}
+
+func (r *simpleRunner) StatusAll() []runner.ProcessStatus {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	statuses := []runner.ProcessStatus{}
+	for _, status := range r.runs {
+		statuses = append(statuses, status)
+	}
+	return statuses
+}
+
+func (r *simpleRunner) Abort(run runner.RunId) runner.ProcessStatus {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	result, ok := r.running[run]
+	if !ok {
+		return runner.ErrorStatus(run, fmt.Errorf("could not find abortable run: %v", run))
 	}
-	return result, nil
+	result.Abort()
+	delete(r.running, run)
+	r.runs[run] = runner.AbortStatus(run)
+	return r.runs[run]
 }
 
-func (r *simpleRunner) markFinished(p execer.Process, runId runner.RunId, status execer.ProcessStatus) {
+func (r *simpleRunner) Erase(run runner.RunId) {
+	// Best effort is fine here.
+	if result, ok := r.runs[run]; ok && result.State.IsDone() {
+		delete(r.runs, run)
+	}
+}
+
+func babysit(p execer.Process, runId runner.RunId, r *simpleRunner, timeout time.Duration) {
+	timedout := false
+	done := make(chan interface{})
+	if timeout > 0 {
+		go func() {
+			select {
+			case <-time.NewTimer(timeout).C:
+				timedout = true
+				r.Abort(runId)
+			case <-done:
+				return
+			}
+		}()
+	}
+
+	status := p.Wait()
+	if timeout > 0 {
+		done <- nil
+	}
+
 	r.mu.Lock()
 	defer r.mu.Unlock()
-	r.running = nil
+
+	// Check if we already marked this run as done (i.e. in Abort())
+	// Further, check if we aborted due to a timeout.
+	delete(r.running, runId)
+	if r.runs[runId].State.IsDone() {
```

Comment:
It it's already done, shouldn't we not change its status? Like, if it were aborted, it shouldn't later change to timed out?